Add failing test fixture for AddArrayReturnDocTypeRector – not compatible with generic parameters

jkuchar opened the pull request:
# Failing Test for AddArrayReturnDocTypeRector

Based on https://getrector.org/demo/d79b7b37-2939-4f8c-99f8-dc1602888273

AddArrayReturnDocTypeRector should keep generic `@return` php doc intact. As current doc block is more specific that what has been inferred.

jkuchar committed:
Add failing test fixture for AddArrayReturnDocTypeRector

# Failing Test for AddArrayReturnDocTypeRector

Based on https://getrector.org/demo/d79b7b37-2939-4f8c-99f8-dc1602888273

TomasVotruba commented:
When I dump the return type, it looks like PHPStan is already providing wrong type:

```php
PHPStan\Type\ArrayType #73017
   keyType: PHPStan\Type\MixedType #73016
   |  isExplicitMixed: false
   |  subtractedType: null
   itemType: PHPStan\Type\ObjectType #73018
   |  className: 'Rector\TypeDeclaration\Tests\Rector\ClassMethod\AddArrayReturnDocTypeRector\Fixture\T'
   |  subtractedType: null
   |  classReflection: null
   |  genericObjectType: null
   |  superTypes: array (0)
```

Could you confirm on phpstan.org that `@template` syntax is resolved by PHPStan correctly?

TomasVotruba commented:
NameScope doesn't show any templates either :/ 

```php
PHPStan\Analyser\NameScope #74786
   namespace: 'Rector\TypeDeclaration\Tests\Rector\ClassMethod\AddArrayReturnDocTypeRector\Fixture'
   uses: array (0)
   className: 'Rector\TypeDeclaration\Tests\Rector\ClassMethod\AddArrayReturnDocTypeRector\Fixture\DemoFile'
   functionName: null
   templateTypeMap: PHPStan\Type\Generic\TemplateTypeMap #73223
   |  types: array (0)
```

TomasVotruba commented:
Got a lead :) we're not covering templates in Rector NameScope factory

jkuchar commented:
I have updated code sample a bit, but it looks like phpstan resolves this correctly:

- phpstan: https://phpstan.org/r/9caf1210-400b-4580-abb4-e657aac11939
- updated rector: https://getrector.org/demo/ed1e4b25-5092-4f08-91fd-1b309624fac9

jkuchar committed:
Update base_model.php.inc

TomasVotruba commented:
The missing `@template` support in Rector was the problem.
Is fixed here https://github.com/rectorphp/rector/pull/5019
